_posts/2019-04-16-preservation.md

```diff
@@ -0,0 +1,358 @@
+      (Coldline) with constant access times, whereas for access we would have had
+      to wait a few hours for Amazon Glacier.
+   1. Files in GCS hold their own checksums as a property of the file, and the
+      uploaded metadata also contains our own calculated checksums.
+
+### Test Cases
+
+The following test cases are largely academic. We haven't done each of these,
+but they've helped to think through what it would take to implement each in the
+above options.
+
+1. Preserve a Book with N pages.
+1. Delete a full book
+1. Delete a page from a book which is already preserved.
+1. Add a page to a book which is already preserved.
+1. Preserve a Book with Books which have Pages (A multi-volume work)
+
+### Preserve a Book with N Pages
+
+#### BagIt
+
+1. Create a base directory named `<book-id>`
+1. Create a `bagit.txt` in the base directory.
+1. Optionally create a `bag-info.txt`
+1. Iterate through each page and create a file in `<book-id>/data` named
+   `<file-id>-<file-name>.<file-extension>` with the binary data. Store the
+   metadata in `<book-id>/metadata/<file-id>.json`.
+1. Either when done, or as each file is added, update `manifest-md5.json` and
+   `tagmanifest-md5.json` with the checksums for each file in `data` and
+   `metadata`.
+1. Optionally update `bag-info.txt` to include the `Payload-Oxum` and `Bag-Size`
+
+**Pros**:
+
+1. We have a bag that people can look at and reasonably understand.
+1. External systems may possible be able to ingest the material, but in practice
+   this is pretty unlikely - bag profiles vary substantially from vendor to
+   vendor.
+
+**Cons**:
+
+1. The manifest files act as a pessimistic lock on the system.
+   * If we have to wait until everything is "done" uploading to update those
+      manifests, then we need to have some way to tell when we're "done." To do
+      this, in Ruby at least, either the uploads must be done sequentially on a
+      single system or some sort of external locking mechanism must be used to
+      keep track of current upload state.
+   * If instead of waiting we simply update the manifest when a file is
+      uploaded, then it means we need to have the ability to lock the file itself
+      in place to avoid writes from multiple processes running into each other.
+1. We have to prevent future changes to a child from persisting to preservation
+   storage until this large change is done. Otherwise, the manifest will be
+   corrupted by a race condition:
+   1. Preservation starts. Pages start to persist into a bag.
+   2. Page gets modified out of band (label changed or page switched)
+   3. Page's changes get sent to be preserved.
+   4. Page gets preserved in its spot in the bag, updating manifest.
+   5. Initial preservation ends. Updates manifest, but has out-of-date checksum
+      information.
+
+#### OCFL
+
+Implementation notes for OCFL recommend things such as pair-trees for storing
```

Comment:
It says you can do this not that you need to do this. We know it's a use case many currently have so we addressed it.

We'll clear this up that it's one potential way of storing in the implementation notes.

Reply:
:+1: Apologies, I'll change the wording to "Implementation notes for OCFL provides options for things such as...", is that a better representation?